### ISSUE TITLE
Component based filters

### ISSUE THREAD

**Issue by Rahel-A** [variar/klogg#191]
I have a really complex regex filter in order to hide noise and look at my relevant matches, and was wondering if there was a way of simplifying it through GUI instead of regex.

Here's an example of a match filter that I use:
```-ComponentX.+<.+T:(?!ignoreThreadA|ignoreThreadB).+M:(?!ignoreModuleA|ignoreModuleB).+F:(?!ignoreFunctionA|ignoreFunctionB).+>```

I'd be happy to work on this functionality if it's possible to implement this in GUI. For example I'd like to be able to check tickboxes to view various modules and components.

**Comment by variar** [comment]
I've been thinking about this for a while. Looks like we need several things to make it work:
 - some parser to extract "selectable" parts from each line and store them in the index. Ideally with user-configurable rules
 - a widget with some kind of checkboxes to include/exclude lines based on extracted features
 - modified search routine that will use index of per-line features and configuration from the widget to filter lines.

There's a much simpler 'alternative' -- an additional textbox to enter exclude regex. That way lines matching exclude regex will not appear in filtered view. Adding selected text to exclude filter via context menu could save some typing.

**Comment by jh12z** [comment]
I think it'd be very helpful to mix highlighting, filter (include/exclude) and easy activation/deactivation with a checkbox.
Something like Notepad++ [Analyse Plugin](https://sourceforge.net/projects/analyseplugin/), pity that one only works in Windows.

**Comment by dickysyai** [comment]
hello anyone can tell me how to clear history search keyword in klogg ?

**Comment by twardakm** [comment]
It's not exactly solution for the requested feature, but I had similar problem on my own, and implemented checkboxes with predefined filters.

Would you be interested in merging that feature? For now I have only hardcoded filters in my source files, but I was thinking of using configuration file of klogg to save those.

![Predefined filters](https://user-images.githubusercontent.com/5730030/101357343-98d24180-3899-11eb-9f15-3b7adbb35446.png)

**Comment by variar** [comment]
@twardakm yes this feature looks promising. In my workflow I use autocomplete and history in search combobox. Configuring some predefined searches could help

**Comment by twardakm** [comment]
great! @variar I will prepare a commit in the coming days (or weeks depending on my workload). I am thinking about reading filters from klogg.conf file.

**Comment by twardakm** [comment]
I published first commit here: https://github.com/twardakm/klogg/commit/4442f33954b725ecb2d442e147a79c077eb05073

Example klogg.conf looks like this:

```
[PredefinedFiltersCollection]
filters\1\filter=.*kernel.*
filters\1\name=kernel
filters\2\filter=.*spotify.*|snap.*
filters\2\name=spotify
filters\3\filter=[0-9]+
filters\3\name=number
filters\4\filter=firefox|chrome|vivaldi
filters\4\name=browser
filters\5\filter=.*kde.*|.*qt.*
filters\5\name=kde
filters\size=5
version=1
```

Another commit with dialog window for setting up filters is coming soon.

**Comment by variar** [comment]
Looks promising. I think that one simple way to create a predefined filter might be a a context menu option for search text edit that allows to create named filter from current regex.

**Comment by twardakm** [comment]
@variar , I didn't thought about that at the beginning, but it's a nice idea. For now I implemented dialog to edit/create filters and I think dialog is just better to manage bigger collection of filters (which is certainly my case at work, where I have ~100 of filters), but context menu option would be great for quick saving current filter.

If you like this idea, let me know, I create pull request.

![Predefined filters](https://user-images.githubusercontent.com/5730030/103307418-42779f00-4a10-11eb-89be-77ed44df4b20.png)

**Comment by variar** [comment]
@twardakm yes, such dialog is needed to manage large collection of filters. I'm looking forward to seeing a pull request :) 

This dialog can also be used to export/import filters to share within a team (like in highlighters), but I think that can be added later.

**Comment by twardakm** [comment]
@variar , now as we have it merged, do you want to create separate issue for export/import feature? I will be willing to implement it as well.

**Comment by variar** [comment]
@twardakm lets move predefined filters to issue #243
